applications/rollout_ensemble.py

```diff
@@ -0,0 +1,626 @@
+            nlon=conf["model"]["image_width"],
+            device=device,
+        )
+
+    # Bred Vector Integration
+    # Add the following variables at the start of your code for bred vector parameters
+    perturbation_size = 0.01   # Magnitude of the initial perturbation
+    epsilon = 0.5             # Rescaling factor for bred vectors
+    num_cycles = 3            # Number of bred vector cycles
+    results = []              # Results storage for ensemble predictions
+    ensemble_members = []     # To store each ensemble forecast
+
+    # Rollout
+    with torch.no_grad():
+        forecast_count = 0
+        results = []
+        save_datetimes = [0] * len(forecasts)
+        ensemble_members = []
+
+        # model inference loop
+        for k, batch in enumerate(data_loader):
+            batch_size = batch["datetime"].shape[0]
+            forecast_step = batch["forecast_step"].item()
+
+            # Set up dictionaries for metrics results
+            metrics_results = [defaultdict(list) for _ in range(batch_size)]
+
+            # Process initial conditions and reshape data
+            if forecast_step == 1:
+                # Handle datetime information for the batch
+                date_times = batch["datetime"].cpu().numpy()
+                init_datetimes = [datetime.utcfromtimestamp(dt) for dt in date_times]
+                init_datetime_strs = [dt.strftime("%Y-%m-%dT%HZ") for dt in init_datetimes]
+                save_datetimes[forecast_count:forecast_count+batch_size] = init_datetime_strs
+
+                # Reshape input data for full batch
+                if "x_surf" in batch:
+                    x_batch = concat_and_reshape(
+                        batch["x"],
+                        batch["x_surf"]
+                    ).to(device).float()
+                else:
+                    x_batch = reshape_only(batch["x"]).to(device).float()
+
+                # Add forcing and static variables
+                if "x_forcing_static" in batch:
+                    x_forcing_batch = batch["x_forcing_static"].to(device).permute(0, 2, 1, 3, 4).float()
+                    x_batch = torch.cat((x_batch, x_forcing_batch), dim=1)
+
+                # Load y-truth
+                if "y_surf" in batch:
+                    y_batch = concat_and_reshape(
+                        batch["y"],
+                        batch["y_surf"]
+                    ).to(device).float()
+                else:
+                    y_batch = reshape_only(batch["y"]).to(device).float()
+
+                if "y_diag" in batch:
+                    y_diag_batch = batch["y_diag"].to(device).permute(0, 2, 1, 3, 4)
+                    y_batch = torch.cat((y_batch, y_diag_batch), dim=1).to(device).float()
+
+                # Clamp if needed
+                if flag_clamp:
+                    x_batch = torch.clamp(x_batch, min=clamp_min, max=clamp_max)
+
+                # Generate bred vectors
+                bred_vectors = []
```

Comment:
Can we move the bred vector functionality to a function in a credit module? It would be more reproducible and testable that way.

Reply:
I moved the code to a new directory called credit.ensemble.